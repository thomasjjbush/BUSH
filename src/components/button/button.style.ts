import styled from 'styled-components';

export const StyledButton = styled.button`
    align-self: flex-start;
    align-items: center;
    appearance: none;
    border-radius: 5px;
    cursor: pointer;
    display: flex;
    font-size: 18px;
    font-weight: bold;
    justify-content-space-between;
    padding: 10px 20px;
    margin: 20px 0 0 0;

    img {
        margin-left: 10px;
    }
`;

export const StyledPrimaryButton = styled(StyledButton)`
    background-color: ${(props): string => props.theme.colors.brand};
    border: 0;
    color: white;
`;

export const StyledSecondaryButton = styled(StyledButton)`
    background-color: transparent;
    border: solid 2px ${(props): string => props.theme.colors.brand};
    color: ${(props): string => props.theme.colors.text};
`;
