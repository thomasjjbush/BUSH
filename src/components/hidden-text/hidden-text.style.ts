import styled from 'styled-components';

export const StyledText = styled.p<{ clamped: boolean; lines: number }>`
    display: ${(props): string => (props.clamped ? '-webkit-box' : 'auto')};
    overflow: hidden;
    text-overflow: ellipsis;
    -webkit-line-clamp: ${(props): number => props.lines};
    -webkit-box-orient: vertical;
`;
