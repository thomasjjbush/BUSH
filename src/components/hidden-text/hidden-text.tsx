import React, { FunctionComponent, ReactElement, useCallback, useEffect, useRef, useState } from 'react';
import { HiddenTextProps as Props } from '../../types';
import { Button } from '../';
import { StyledText } from './hidden-text.style';

export const HiddenText: FunctionComponent<Props> = ({ children, lines }: Props): ReactElement => {
    const ref = useRef(null);
    const [clamped, setClamped] = useState(true);
    const [refLoaded, setRefLoaded] = useState(false);

    const unClamp = useCallback(() => setClamped(false), []);

    useEffect(() => {
        if (ref && !refLoaded) {
            setRefLoaded(true);
        }
    }, [ref]);

    return (
        <div>
            <StyledText clamped={clamped} lines={lines} ref={ref}>
                {children}
            </StyledText>
            {clamped && ref?.current?.scrollHeight > ref?.current?.clientHeight && (
                <Button onClick={unClamp}>View more</Button>
            )}
        </div>
    );
};
