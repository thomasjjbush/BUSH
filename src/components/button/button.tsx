import React, { FunctionComponent, ReactElement } from 'react';
import { ButtonProps as Props, Buttons } from '../../types';
import { StyledPrimaryButton, StyledSecondaryButton } from './button.style';

export const Button: FunctionComponent<Props> = (props: Props): ReactElement => {
    if (!props.onClick) return null;
    return props.design === Buttons.PRIMARY ? (
        <StyledPrimaryButton {...props} type="button">
            {props.children}
        </StyledPrimaryButton>
    ) : (
        <StyledSecondaryButton {...props} type="button">
            {props.children}
        </StyledSecondaryButton>
    );
};

Button.displayName = 'Button';
