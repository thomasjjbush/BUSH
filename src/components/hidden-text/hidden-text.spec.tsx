import React, { MutableRefObject } from 'react';
import { shallow } from 'enzyme';
import { HiddenText } from './hidden-text';
import { Button } from '../button/button';

function mockRef(clientHeight: number, scrollHeight: number): MutableRefObject<Partial<HTMLParagraphElement>> {
    return { current: { clientHeight, scrollHeight } };
}

describe('HiddenText', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('should match snapshot when not scrollable', () => {
        expect(shallow(<HiddenText lines={2}>My text</HiddenText>)).toMatchSnapshot();
    });

    it('should match snapshot when scrollable', () => {
        jest.spyOn(React, 'useRef').mockReturnValueOnce(mockRef(100, 200));
        expect(shallow(<HiddenText lines={2}>My text</HiddenText>)).toMatchSnapshot();
    });

    it('should unclamp text and hide button on click', () => {
        jest.spyOn(React, 'useRef').mockReturnValueOnce(mockRef(100, 200));
        const wrapper = shallow(<HiddenText lines={2}>My text</HiddenText>);

        wrapper.find(Button).prop('onClick')();

        expect(wrapper.find(Button).exists()).toBe(false);
        expect(wrapper).toMatchSnapshot();
    });
});
