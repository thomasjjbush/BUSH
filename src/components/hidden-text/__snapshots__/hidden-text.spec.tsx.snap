// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HiddenText should match snapshot when not scrollable 1`] = `
<div>
  <styled.p
    clamped={true}
    lines={2}
    styles="
    display: -webkit-box;
    overflow: hidden;
    text-overflow: ellipsis;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
"
  >
    My text
  </styled.p>
</div>
`;

exports[`HiddenText should match snapshot when scrollable 1`] = `
<div>
  <styled.p
    clamped={true}
    lines={2}
    styles="
    display: -webkit-box;
    overflow: hidden;
    text-overflow: ellipsis;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
"
  >
    My text
  </styled.p>
  <Button
    onClick={[Function]}
  >
    View more
  </Button>
</div>
`;

exports[`HiddenText should unclamp text and hide button on click 1`] = `
<div>
  <styled.p
    clamped={false}
    lines={2}
    styles="
    display: auto;
    overflow: hidden;
    text-overflow: ellipsis;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
"
  >
    My text
  </styled.p>
</div>
`;
